Ingest correctly handle upsert operations and drop processors together

joegallo opened the pull request:
Closes #36746

`new ShardId(String index, ...)` checks that the `String` arguments it receives are not `null`. When an upsert and a `drop` processor meet, the thing returned from `getIndexWriteRequest(...)` won't have the index populated, so we end up passing `null` where it can't go.

But we don't need to call `getIndexWriteRequest(...)` at all! The `docWriteRequest` argument to it already has everything we need **or so I think** (feel free to dis-abuse me of this notion).

~WIP because I have cleared all out my test changes and added some extra asserts. Once I get a green run from Jenkins on that, I'll revert the revert and get my new tests back (but lose the extra asserts). After that I might still add a test or two.~

joegallo committed:
Use Objects.requireNonNullElse for this

joegallo committed:
Use the docWriteRequest for these values

joegallo committed:
Whitespace

joegallo committed:
More whitespace

joegallo committed:
Add a failing test of drop+upsert

joegallo committed:
Tighten up the failure side, too

It doesn't invoke new ShardId(...) which is where the exception was,
but it's still nice to report this information if we have it (which we
do!).

joegallo committed:
Rename this constant

joegallo committed:
Swap the order of these methods

Failure is so much more normal/expected than dropping, so let's have
that one first. Also it has a better explanation of what's going on
and why, which I like having earlier (but don't want to duplicate).

joegallo committed:
Add another failing test

joegallo committed:
Tidy this up a bit

Increasing the symmetry between markItemAsFailed and
markItemAsDropped.

joegallo committed:
Rely on static imports for these

joegallo committed:
Temporarily blow away the new tests and add asserts

elasticsearchmachine commented:
Hi @joegallo, I've created a changelog YAML for you.

joegallo committed:
Update docs/changelog/104585.yaml

joegallo commented:
<img width="862" alt="Screenshot 2024-01-20 at 12 44 19 PM" src="https://github.com/elastic/elasticsearch/assets/187034/025f9898-7e2c-4a3d-8194-c4c669648015">

15/15, so I'm adding the tests back now.

joegallo committed:
Revert "Temporarily blow away the new tests and add asserts"

This reverts commit 29ce3c652d1631848a5b1cf22906ada40ed28e80.

joegallo committed:
Merge branch 'main' into ingest-handle-upsert-with-drop

masseyke commented:
Seems like a good idea. One test that would be good to have (and this might be one you already have in mind) is a non-bulk upsert since the issue reports a different stack trace for non-bulk. I think the current tests only cover bulk, unless I missed something.

joegallo commented:
>  One test that would be good to have [...]

Yes! That's the precisely the test I have in mind.

joegallo committed:
Add a test for fail with _update

This requires that we have a default_pipeline in place, so I've also
extended the cleanup to include deleting the index -- otherwise if
something goes wrong the cleanup can fail because you can't drop a
pipeline if there's an index using it.

joegallo committed:
Widen this type

The response will be an UpdateResponse (not an IndexResponse) when a
document is dropped on an upsert, both of those are a
DocWriteResponse.

joegallo committed:
Add a test for drop with _update

This requires that we have a default_pipeline in place, so I've also
extended the cleanup to include deleting the index -- otherwise if
something goes wrong the cleanup can fail because you can't drop a
pipeline if there's an index using it.

joegallo committed:
Skip versions where the bug hasn't been fixed

joegallo committed:
Merge branch 'main' into ingest-handle-upsert-with-drop

joegallo commented:
@elasticmachine run elasticsearch-ci/packaging-tests-windows-sample

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

joegallo commented:
A bunch of the lines of code changed here were just whitespace twiddling in the tests, I'd recommend reviewing this PR with [`?w=1`](https://github.com/elastic/elasticsearch/pull/104585/files?w=1) to ignore all that.

masseyke commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -1110,35 +1109,31 @@ private void assertResponsesAreCorrect(BulkItemResponse[] bulkResponses, BulkIte
+            final DocWriteRequest<?> docWriteRequest = bulkRequest.requests().get(slot);
+            final String id = Objects.requireNonNullElse(docWriteRequest.id(), DROPPED_OR_FAILED_ITEM_WITH_AUTO_GENERATED_ID);
+            failedSlots.set(slot);
+            UpdateResponse dropped = new UpdateResponse(
```
Why is this an UpdateResponse? I see that it was previously, but it leads to some weird things, like getting an UpdateResponse from an IndexRequest (which you also had to account for in this PR).

joegallo replied:
I don't think IndexResponse lets us express anything except `CREATED` or `UPDATED`:

```
    private static Result assertCreatedOrUpdated(Result result) {
        assert result == Result.CREATED || result == Result.UPDATED;
        return result;
    }
```

But yeah, it is inconvenient (and also ancient!).

masseyke reviewed (approved):
LGTM
